### ISSUE TITLE
Passphrase Functionality Does Not Work in trezor-gpg and Loops Indefinitely 

### ISSUE THREAD

**Issue by TreyFiello** [romanz/trezor-agent#200]
Looking at the fix you put in for https://github.com/romanz/trezor-agent/issues/164 it looks like TREZOR_PASSPHRASE is no longer used? I am wondering if I am using the correct procedure?

1. I enabled passphrase in advanced settings in the wallet.
2. I initially installed trezor-agent via pip, so I attempted to upgrade to the latest github version and remove the old one: 
`pip-autoremove trezor-agent -y`
3. I installed github version:
`git clone https://github.com/romanz/trezor-agent` 
`pip install --user -e trezor-agent/agents/trezor`
```
$ trezor-agent --version
trezor-agent=0.9.2 libagent=0.9.8
```

4. On OSX, I open terminal and type:
`trezor-gpg init myemail@email.com`

5. At which point after making it past the PIN entry, it just repeats "Passphrase required" and "Confirm your Passphrase" indefinitely. After 2 sets of repetition on the Trezor it asks me to confirm signing, but then just goes right back to asking for my Passphrase again.

Output looks like this:
```
$ trezor-gpg init myemail@email.com
2018-02-19 19:46:20,932 WARNING      This GPG tool is still in EXPERIMENTAL mode, so please note that the API and features may change without backwards compatibility! [__init__.py:117]
2018-02-19 19:46:20,965 WARNING      NOTE: in order to re-generate the exact same GPG key later, run this command with "--time=1519094780" commandline flag (to set the timestamp of the GPG key manually). [__init__.py:35]
Use the numeric keypad to describe number positions. The layout is:
    7 8 9
    4 5 6
    1 2 3
Please enter current PIN:

/Users/test/Library/Python/2.7/lib/python/site-packages/unidecode/__init__.py:46: RuntimeWarning: Argument <type 'str'> is not an unicode object. Passing an encoded string will likely have unexpected results.
  _warn_if_not_unicode(string)
Passphrase required:

Confirm your Passphrase:

Passphrase required:

Confirm your Passphrase:

Passphrase required:

Confirm your Passphrase:

Passphrase required:

Confirm your Passphrase:


```

**Comment by romanz** [comment]
Thanks for reporting this issue!

**Comment by romanz** [comment]
Yes, you're right - it doesn't make any sense to ask the user for the same password again and again.
The passphrase should be cached during the installation.

**Comment by romanz** [comment]
TREZOR's Python library now supports `PASSPHRASE` environment variable:
https://github.com/trezor/python-trezor/blame/master/trezorlib/client.py#L284-L287

Please try running (with the latest commit):
```
$ PASSPHRASE="abc" trezor-gpg init myemail@email.com -vv
```

**Comment by TreyFiello** [comment]
@romanz almost working I'm still getting a single dialogue box -- when using:
`PASSPHRASE="abc" trezor-gpg init myemail@email.com -vv`

Key creation eventually succeeds but an input dialogue box pops up (outside the terminal) asking for my passphrase. Maybe there is still a location in the code that is not reading the environmental variable?

The flow was: 
- ran above command
- asked for me to sign on Trezor - hit yes
- asked for me to sign on Trezor - hit yes
- pop up GUI input box asking for passphrase - entered "abc"
- new key was added


```
... SNIP ...
2018-02-20 09:40:23,851 DEBUG        "<gpg://myemail@email.com|nist256p1>" signing 'V\x9c\x13m\x8aT\xac\xdc\xa1H\xfed\x17\xe4\xf9\xb5\x98\xd7CU\x14W\x02Y\xd3\xb2\x9aRi\x1d_\x9f' (nist256p1) on Trezor [trezor.py:153]
Passphrase required. Using PASSPHRASE environment variable.
2018-02-20 09:40:26,995 DEBUG        result: <SignedIdentity: {'public_key': bytearray(b'\x03a\x01w\x90\xd7\x9f\xa7B\xbft\xc4e\xe8\xc8K\x12\x0c\x98\xcb<\xafm\xf0\xc8\xf4_HVc\xf5\xc0|'), 'signature': bytearray(b'\x00\xda\xbc"\x1eQ\xe2\x1e(\xee\xf0\xf8\xb5\xe4\x90P;]\x1cN\x9a\xb5\xbb\xa0-\xef\n\xc6\xf7\x1el\xf8\x0f\x7f\xd1\xbe\xb0\x9c\x05!\xf0/\x88\xab\x1e\xffK\xcc2C\x1eR\xbd\x87\xbf\xb8\x9akMo\xaa\xb8\xa9\x8b\xda')}> [trezor.py:160]
2018-02-20 09:40:26,996 DEBUG        setting /Users/test/.gnupg/trezor/pubkey.asc contents:
-----BEGIN PGP PUBLIC KEY BLOCK-----
Version: GnuPG v2

mFIEWoxPahMIKoZIzj0DAQcCAwRhAXeQ15+nQr90xGXoyEsSDJjLPK9t8Mj0X0hW
Y/XAfB8zkrsGs8A1+/4t8LFZwsSBvrAOlpporEXWCcgxZhpXtBFteWVtYWlsQGVt
YWlsLmNvbYiABBMTCAAcBQJajE9qAgsJAhsDBBUICQoEFgIDAQIXgAIeAQAWCRBA
XaDofeDf9AsaVFJFWk9SLUdQR6EjAP4lG/clTrRx8/DdTLcwVMbrRZ0IimEk6Tpq
rscwa/0QjAD+N+oyY5qQgsfdbiJDExhasBd6DgCcONcSlxsSLnMNzui4VgRajE9q
EggqhkjOPQMBBwIDBCLL0BYB4pKL9yYVmILql3eeHSzUfwXHDeNm9IL7sFVhK8Ji
8iyT1GXnncc0NzAspElq2QTctnUicUHWewcKI3QDAQgHiG0EGBMIAAkFAlqMT2oC
GwwAFgkQQF2g6H3g3/QLGlRSRVpPUi1HUEdWnAEA2rwiHlHiHiju8Pi15JBQO10c
Tpq1u6At7wrG9x5s+A8A/3/RvrCcBSHwL4irHv9LzDJDHlK9h7+4mmtNb6q4qYva
=qd08
-----END PGP PUBLIC KEY BLOCK-----
 [__init__.py:109]
2018-02-20 09:40:26,997 DEBUG        run: [u'/usr/local/gnupg-2.1/bin/gpg', '--homedir', '/Users/test/.gnupg/trezor', '--quiet', '--import', '/Users/test/.gnupg/trezor/pubkey.asc'] [__init__.py:102]

```
**<PROMPTED_BY_DIALOGUE_BOX_HERE>** - If I enter "abc" it completes normally:
```
2018-02-20 09:42:25,693 DEBUG        run: [u'/usr/local/gnupg-2.1/bin/gpg', '--homedir', '/Users/test/.gnupg/trezor', '--list-public-keys', '--with-fingerprint', '--with-colons'] [__init__.py:94]
2018-02-20 09:42:25,701 DEBUG        out: u'tru::1:1519144827:0:3:1:5\npub:-:256:19:405DA0E87DE0DFF4:1519144810:::-:::scESC:::::nistp256:::0:\nfpr:::::::::64DC5F7A00F7B3C0689E22EA405DA0E87DE0DFF4:\nuid:-::::1519144810::5D6058EF46D27545F8248D3750C57247ABF3E207::myemail@email.com::::::::::0:\nsub:-:256:18:CE9FD8BCEDA87798:1519144810::::::e:::::nistp256::\nfpr:::::::::395E139B9CF5D45369842401CE9FD8BCEDA87798:\n' [__init__.py:96]
2018-02-20 09:42:25,702 DEBUG        setting /Users/test/.gnupg/trezor/ownertrust.txt contents:
64DC5F7A00F7B3C0689E22EA405DA0E87DE0DFF4:6
 [__init__.py:109]
2018-02-20 09:42:25,702 DEBUG        run: [u'/usr/local/gnupg-2.1/bin/gpg', '--homedir', '/Users/test/.gnupg/trezor', '--import-ownertrust', '/Users/test/.gnupg/trezor/ownertrust.txt'] [__init__.py:102]
gpg: inserting ownertrust of 6
2018-02-20 09:42:25,710 DEBUG        run: [u'/usr/local/gnupg-2.1/bin/gpg', '--list-secret-keys'] {'GNUPGHOME': '/Users/test/.gnupg/trezor'} [__init__.py:102]
gpg: checking the trustdb
gpg: marginals needed: 3  completes needed: 1  trust model: pgp
gpg: depth: 0  valid:   1  signed:   0  trust: 0-, 0q, 0n, 0m, 0f, 1u
/Users/test/.gnupg/trezor/pubring.kbx
---------------------------------------
sec   nistp256 2018-02-20 [SC]
      64DC5F7A00F7B3C0689E22EA405DA0E87DE0DFF4
uid           [ultimate] myemail@email.com
ssb   nistp256 2018-02-20 [E]
```

Also, when using the newly created key I am also getting 2 dialogue boxes prompting for a passphrase when running a gpg2 decryption command:
`$ PASSPHRASE="abc" gpg2 -d test.txt.gpg`

I assume we want 0 dialogue boxes in all the above cases? It would be very nice to be able to script the passphrase in to all these utilities, for example this ability would prevent someone that saw me enter my Trezor pin from being able to pick up ONLY my Trezor, connect it to a new computer and steal my SSH/GPG keys. They would also need access to my computer to know the passphrase.

**Comment by romanz** [comment]
I can suggest adding the following command to the `~/.gnupg/trezor/run-agent.sh` file:
```
#!/bin/sh
export PATH=<snip>
export PASSPHRASE='abc'
/home/roman/.local/bin/trezor-gpg-agent $*
```

This script is run by GnuPG, so it can set environment variables for the `trezor-gpg-agent` tool.
Please make sure that the `run-agent.sh` script is readable only by you:
```
$ ls -l ~/.gnupg/trezor/run-agent.sh 
-rwx------ 1 roman roman 281 Feb 20 21:19 /home/roman/.gnupg/trezor/run-agent.sh
```

**Comment by TreyFiello** [comment]
@romanz tried the run-agent.sh edit you supplied above. Did not seem to make any difference unfortunately.

After the edit, I did:
`$ gpg2 -d test.txt.gpg`

I was prompted for a passphrase **twice** with the input box.

I then tried:
```
$ export PASSPHRASE="abc"
$ gpg2 -d test.txt.gpg
```

And was prompted only **once** to enter my passphrase.

**Comment by romanz** [comment]
Please try to restart the `trezor-gpg-agent` process, to make sure the running process is using the correct environment:
```
$ sed 's/\x0/\n/g' /proc/`pgrep -f trezor-gpg-agent`/environ | grep PASSPHRASE
```

**Comment by TreyFiello** [comment]
@romanz completely restarted my computer. Confirmed run-agent.sh was still correct.

Entered:  ``$ gpg2 -d test.txt.gpg`` was prompted for pin, then passphrase **2 times**.

Then, tried:
```
$ export PASSPHRASE="abc"
$ gpg2 -d test.txt.gpg
```
And was prompted **1 time** for my password.

Doesn't look like OSX doesn't use procfs:
$ ls /proc
ls: /proc: No such file or directory

**Comment by TreyFiello** [comment]
@romanz one idea instead of using environmental variables would be to consider saving the PASSPHRASE to a file that is encrypted by the PIN. I think this would be even MORE secure than using environmental variables, and certainly more secure than just writing the passphrase to disk.

This could be a configuration option of trezor-gpg:

`trezor-gpg init --cache-passphrase`

After user inputs PIN for the first time, an encrypted-with-the-pin cache file is saved to:
/home/user/.local/bin/trezor-passphrase-cache

If the file exists, after PIN entered, trezor-gpg automatically decrypts and tries it before prompting the user to enter a passphrase.

**Comment by romanz** [comment]
Regarding the different number of passphrase prompts:
The first prompt is for getting the public key (which is later is getting cached at the GPG agent).
The second prompt is for the actual signature operation.

Thinking about it, I guess that the correct way for solving this, is to cache the passphrase at the `trezor-gpg-agent` process (since it's started once, and runs in the background).

**Comment by romanz** [comment]
Please try the latest commit - it should prompt once for the password during the initialization, and once again before starting the `trezor-gpg-agent` (which happens automatically after the initialization is over).

**Comment by TreyFiello** [comment]
@romanz I can confirm I am prompted twice on key generation now.

However, after key is created, doing something like:

`$ PASSPHRASE="abc" gpg2 -d test.txt.gpg`

**Still prompts me for a passhrase**, which means the ~/.gnupg/trezor/run-agent.sh fix is not working? I confirmed the contents and permissions of run-agent.sh are correct as you prescribed above.

What are your thoughts on the disk based passphrase cache idea (encrypted with the 24-word nmneuic(pin))? Wouldn't that solve our issues trying to pass around the environmental variable and make things more secure in general? And also reduce total passphrase prompting to only once at key creation?

**Comment by romanz** [comment]
Could you please attach the log of `trezor-gpg-agent`? (it should be at `~/.gnupg/trezor/gpg-agent.log`)
Also, please try to run the decryption command multiple times - since the caching needs the first UI-based passphrase prompt to succeed before reusing the passphrase in subsequent attempts.

**Comment by romanz** [comment]
Regarding the encryption of the passphrase - this would require the user to enter his mnemonic/PIN on the computer's keyboard, which may causes it to be compromised (e.g. due to keyloggers).

I'd prefer to keep the design simple - for example, keep the passphrase as an empty string for GPG-related operations, even if it's used for the Bitcoin wallet. This was actually the initial design - but then we've added support for custom passphrases (via environment variables):
https://github.com/romanz/trezor-agent/commit/6672ea9bc44681287192628e8e200a4f302e5291

**Comment by TreyFiello** [comment]
@romanz odd - gpg-agent.log is blank when I clear it after key creation, and then do:

`$ export PASSPHRASE="abc"; gpg2 -d test.txt.gpg`

Makes sense to keep the design simple! Just to clarify, I wasn't suggesting to use a plaintext pin to encrypt, I just assumed there was some way to generate a symmetric encryption key AFTER the user has securely entered their pin --- for example, this is how the Trezor Password Vault works. You enter your pin and then they encrypt each password to your dropbox using some sort of symmetric algorithm.

**Comment by TreyFiello** [comment]
@romanz just FYI -- found this code that creates a symmetric key utilizing the Trezor that one could then use to encrypt the PASSPHRASE file if that is still an option:

https://github.com/trezor/python-trezor/blob/master/tools/encfs_aes_getpass.py#L89-L113

**Comment by romanz** [comment]
> You enter your pin and then they encrypt each password to your dropbox using some sort of symmetric algorithm.

This may work, but when I tried the `encfs_aes_getpass.py` tool, the TREZOR asks for a passphrase (since the device is passphrase-enabled). 

I'd suggest to use a static passphrase (e.g. the default empty one) - since a dynamic/encrypted passphrase will probably be more complicated to understand and use.

**Comment by romanz** [comment]
> gpg-agent.log is blank when I clear it after key creation

Could you run `GNUPGHOME=${HOME}/.gnupg/trezor trezor-gpg-agent -vv` from command-line, and `echo 123 | gpg2 --sign -sa` in another shell?

When running from command-line, `trezor-gpg-agent` should log to stderr.

**Comment by TreyFiello** [comment]
Bummer that the aes_getpass.py prompts you for a passphrase, sorry to have wasted your time :-(

I still have ~/.gnupg/trezor/run-agent.sh setup.

Running the command you asked:
```
$ echo 123 | gpg2 --sign -sa
gpg: Warning: not using 'myemail@email.com' as default key: No secret key
gpg: all values passed to '--default-key' ignored
gpg: keydb_search failed: Broken pipe
gpg: no default secret key: Broken pipe
gpg: signing failed: Broken pipe
```
```
$ gpg2 --list-keys
/Users/test/.gnupg/trezor/pubring.kbx
---------------------------------------
pub   nistp256 2018-02-21 [SC]
      A834A35AFCE8011FCE365A13A60EF3C6B44D86D2
uid           [ultimate] myemail@email.com
sub   nistp256 2018-02-21 [E]
```

I then did `gpg2 -d test.txt.gpg` and here was log output in trezor-gpg-agent (PGP encrypted this, not really sure what is sensitive in here :-)):

```
-----BEGIN PGP MESSAGE-----
Version: Keybase OpenPGP v2.0.76
Comment: https://keybase.io/crypto

wX4DT6XItd1c+h8SAgMEm9DQhzU+22cVQVxaoLjcNZNC4mZs25pHCYjaRK/6VBxc
BHpgx/eYmoO8vL8IjlB+D5fPCx7Hm2GCOrj6Wde3hTAwK0/gyutWGdNO0/mMK64E
YbtGumP3o5IZb5dbwpBL0MEKbLuU/Ch7QVYJA81EvvPS22YBQGNcPAiB6jlZKmOR
AHklVHzPUJxb8E8gOxj1BDFknboUFyal5EJqYY563JKmO3O0qB/S7ahvgacJyPBV
pYxX4/qIsuRFggasJk35Yah47GVlWfb0QLaYbeJ7XGKpxgLX5C8O0AMdegt4uHtN
zYgIVJizjfNLsNgfCGTxQYuzlopv6+YUokxgGmmv4jE93NXDqA1FfUqlSruvazvZ
5gmJYqJ1MkivlA3d13zN12iAgQI41FhilGceJHWnCWPSKQrktd/730j1YfxO4TJx
1BT4/IEbPB9AuZmJiGcZzwWxh7UQnbVVrq+9GaSKk5O8xw8SPwlVxK9TikoDgmMd
fvqw/NTqFkBklps0+tW35RmjJ7nlUfSiHx41xOvLDXx3Kw9v6yuvEANso8S+5UWs
9MeJP0QbdaUwUk+/S2o2Q8sO7FyhUsqHRexQMv9J1kcL/IZAJrmeDG3WhNZCQ21U
TWkCK0KDf6aQ8PagroAiRZPKwLeZ100bL+YSNJCwRVf7VdOdbZ6ZxpHT0W5y9y05
cw9nIG5RNWBSVvaSr2bD8hxJFgU6/gAGGtA5fpVziGIILb7xIJI6J3NnowUfOORo
cAfJHlWKBZLcRlNJMRn2ZAG9Na+/gGZxlecUt6El5mSRndfKFXqIXxvIGkaJS5QX
zpey7FZ6r2Kec/yofCOqz9gefIODwtMegW8NPrqmJbJmI9R/sFcaWMwQlr0Cnz1u
8buc+hq1w0AMTPel+ldI0nbgpbqN3JK3N9TR5kihxD62oMS4hbC5lZEN1m+sXCjs
TbhEU9wouspzj6IOOKYxV/Kuq2obq5+19ruonZHuRqWwgg39rqFi3dcTiDXUKrtd
7Zb8CtR54uOV+8bKnAZ7/rW+HI6AWsbg8MuIj6c8tr/c8+FHaj+hLqGtLrozdT2N
y3d5ejAXYpGr/YHptCoBw8l1Eck1SVSoU7Z5nrfY7rCPpN2pTz6hWzXWTyVTiR29
oeyF/RgoFoPpqHP1zblcRsrm1kuJohw0yqRfef2WnnhLgKdZbsWRBwq+U5X6F8mB
Dp3XKBM01qlp/lsNLcrGNwyYcsQnsgTFwqENaVJuQ0tRmS4NvTfa81q7KO1TMxix
uQJDLXfVqDLon7Kd1pK3n0O4OJhURuoUgzw7+yuSjR3YA6atnU8Q+l6CfBwuSyF7
xWJ+DpWxIwdJFJKN9Stjm7MwCBR5/4LpDz3uWOULck9G/A0O4fr2joNFFpgVOfDQ
nYWlXp+U7GRMQbfMZEF0BkpNcfWJqnCOmPFvPlWI4npyB641STLteeqVDRiULFAK
4IYhfYPmgVn24UGn54Golm39RGLJQ9d0MQTZZ9pfXSC6Qum/pp3LofWbf1haQW81
SVyOZtd+irww1WqeEQ74QUwRy/UFnC6hsG3kwGA1ARfO6uKLEOZMj+8itjGvPF+/
uUhbD+kYrlebvcINjrlDjxGTdHDm9B5OS5Vn+YrvtI3Uo2bEvUvNzWfxxLuo+USF
z/uUaN+Q7V/ny0nHL3OpoGa6RYgcNizKY7ijNTlEZN2pMa1eXE4nzd9mFK0aC95P
3q1nlAO93aXFbyeFdw0HvChkVjEHRPL9iY+lN5w4qH6wARJX7umZvrK0EU/ZJy6c
PZkkL3jYjW2r9Gs5VONj+Jb+Vt7/p07ZZSkr9vL/lNrkPMsygNBWhF28eqWqiCT9
c2ER5ngI7egZbz6NyJ6C5PuNX1W9nSTnJp6u06mgFoIK5aMXpPyWFYx7l0LV04DR
7awmiU62+dNqsJwbEAjLkhL4NUjx6bDRqEakaYvsupLSb1yc80pCcYHKEhhluo2t
2qxMGhS/KL5ShhRBpnW8QF+zJ52Qa0/UKt+HHXhaDbm39uEJVQuesEpAgcAmLCBm
PsWTV7vuAX/5Z0aTYrqsgSQORX2O0fgJxvRcUBaCFzwUq+khPjFeZWKoJ3D4Q/x2
bAie0bKWbgjOkOkVhwPXj95sBvot+G3dlt27r22mIpkzompNyJE4/b491Zf7giso
kWQD2745krAnh+8U4fWOoVEZHD+sWhRopITDwJi8e22IRtUvZXCIvK5GCONLN9oA
aakUCZELzivTKVzGo6sRwW/3KR8FofGzE8sCiPeiQkLbNB90/8RYuCNM/kG8So5+
rhihKnnO1/662c0T5QuDxlsCXUAR2vIlthIQuw4cMwSq9aR7coR/DVfE6GYYcECU
wMoB1ucjlNKmwBHxf/GSpkUxellU2kbTrTkvPJrAG7gAV920TzxFbYcsrdj+UWaz
IW75zuxgvc34bwBSyGygqaC/QUvi5dD630uxaK+NVj+BzJGvQ2j+L+zgjBuClUUi
uyupuE244QiaxBhWNYjST/e8TmOcmT6w2L0MO0NoS++vEz+Sosrly67rZu2k90W/
7QjDBQjwmxpafryUKR0DgEzd6kxChIDX8mGSGu17u8Km+nlshuDPhdQ+KLDjPKaR
udqrEZK5WCeTTSmj0fOjDMVwvvO5TOFV8O6ro0UDNxsVDXPaKJ1t7l7PNvIthcFk
L0LOXDT61Hfgi1KtDjhJ0bvWl8KfJarZJSMsp0H+zg59TGyFsUG6kDqDJUvSESlS
HtmgRURJACMYa1ex1UDq77u3jt8oRYbvEDOE9DA3EvLHqVidINgFUFFYvv3ecJov
9vr4UYwVQZJpK1OT1JqlVrZlLHa7JESFizcID+KNGBd2SzGkZlKqDrPcYIfFWfcI
gxfJThCofESj2yCqkj/KnBXVXyx/duhWmrFKXH8P+8cSfynMDaS/Q+C/g/mv76Om
SgBX/Il6ziqcWo0KoIhpeNxfSHnbiFpgMFnxyRSNLz0gWS8LejO2EYk4UHMzTa5D
1SOJHI1sfP+IbiFW8F7nd7SIP2cdm01xpnCSEAF55lilFogtkvDm3A08C0zYmtUW
eQ375xlzmta0+RnZflSe2bCx/W/lY/bNHQQa22uVvMFEBZ80K6MqDOHRIS2X+psy
d2V55VaDZondzgUAV2sudi7IC5opRdOQyzDvvJLQ2qziFxujvJJjZqGsLZE4+rOO
SxB0Jz2hRAWZdCdO4tYtyFgXPEthC2tA3QCNXPYpDWJUZi5iJik68yYtDS+XHzIG
5593uTcAo1Jr5D6zVY6Wh0DWl9f2Vc4s5ppmXW44m7LuDy6/GU7I0NQZE4cwfQs7
LuH7jUT3qME/sgC7ldb2E1fOiQQsL9X09AVkHSSrAc2wg0+Emi4vYOAdX/eelwcz
ZtcwiqFWR3HfxyBa96KbD5MNxY6odIsQ2qsX51OQ8WIIp/qja/2pD4VxkaxUz6Pz
td3DJP/7RA4zCd7gykFDvY1reskGOaaD5rMwt/WxF9vOnAPLhgxPw6TI+ug3e+Nd
c0RzbifPYY28dg+kHABPM/mxPwVpZ3nbUpi4tL2jRDd22H6vy1ojxUoxHJ8PspfR
5BdsJs8Vml6Ubw9zaliM7bMuxoinbYhoZ01+NclQYjPI0wKKB24hb6OBSXI0Cpmo
CQ/tJhARTsF21QuZ1a4sUBjnLO71nkXOEVusFfsV5/dNe2i2tbkpz56TXp8M1msk
iXKZKcwe3GVCXVevrf8v65k6YoRxf1T7aUY1qWMe2k3hrrClfvDRr2iSDEn7ET1q
ri4jhWFd9b/Sm9JzlUCTUuvrwZOvOEIR6nvj/Bp1QWXf6/hLVoW20FsTBLoELzIH
gX3O65UIkLlfFXlNOlSsxQ4Yvk5qxeyXRnTttwR8/FZnXeMv1D7bwUvanXBbJa55
MlmALvPJGohuqLPTS2iNgepN/K9bwvxQ6KJPOCzqeW+kuLbZ6sL4lmMGKOLI3fYD
lRN6rY9+Konvyua0IBdXc591dILljCHnPJU6OnnAMOlbVacbBhk+iY9yctWZxEuv
m6RbpuYGg0esumjRXpl6cuEajlVFWoSYSnFhK52j+u4EvLSXintd51rMIt9Psd0i
eYpIvm0r/PA8cczjJnhfq1syuiyvj0FaACVY0aecHuu3fm18LB2YEeD2RsOGUe5O
T0FKj81MDJaRMRfA/hGZ41M12K61iG5ijCLpv5W3UXvk6khnWae0FHIVE7hDN3c5
ING8prwA6Zq2Lfk4WbQebL22ThDVLJadlOoLG83kPt+EIrGgFY9YlYcVeQ7ElLeE
KeP55Zrgqf3xerQNGagjXuaTFcC2fVv6S34QE2FFlwCcAAb84idqBjZYtqUSLcjh
d1eUwQPyTrXmSd2ILTzB/9oKXLvdrU3hASjxT5Morm9kY+H29uIqhxQ5YXqsTtD2
SaoJppREAImXYTSDIgi8T7KphMOgeP5vlxDzxkaLMSkkXIETZVISNjDFyXLheug+
UCGNJtQDeFDyXsvc8S/vQsWbalh28Jt+jqoELQq1ViVUAWf/pmUV3IWtDAsI+VXI
JC6ptpn22dsKi7NTeSx8LThUqlb3tDegC3FIxSRRG0WyFNxABzAklp/Wve/F7r+9
qIpLNo9S5UEgjyDdmsS6JHWN/NMkhH7lM2nzKyNdYBBFeFWlWzjpjzyGEObszUCt
NKqWSQTgOM8xSDA58WSpqzaVcWGhDxl33TkY6HAMjlXqF+DvX/dK2rQ2b+F00OPZ
fu0+83OEtw2Qq7mJB3NUqT3/ByuK7Y3yug/UDYdYFuPABH3KjVduTcb2daT7j9zH
y88YTz59bUpfS4+dtdWz0nnKdP1CV/Dt7giETjaS9hkH2w4JpiKDX48c5fZ380JG
B8U44C84BPjW1Pn/imOrwrfT5nlMswmtt7ZeBblzUzIQ/N4L5KbKboFeqwDF2qrZ
GfsBtoUSiy4S9Bi+LzWsDk7Z3ZrXeRlfxwdZ5Wf+HoAZgvOycPYBkxOMnFE6oDON
Ob9/nZi2UfJPleGKqKBUUVABYQtlZwWSb6DUadQgiTq5Qw5dhrCwfF6IEcy6h6Ox
SkfMXzg6RYHZFTVzKRjdyvA1EluZXy57Osfy1N9WJF9XselTageqOhVLbiGLdSlU
1EWvkn00GG9UhOy4lGXHn23YtxpDE7C8OG5eFe2TNJvi/G7fyp2AbRBBZUV1bNex
AoHhiWbK1ExNu4Q3CS8a3INMOWsbkFmxVKO2SWSJJ+z2k3+8D+iNI24Xm+UGfbQI
vvCOvJ9fTOogSZ10JWahJlQPvQBp3uTYpnplIv6wedc7XIoqPX12U9pu118Goicn
lFDywcadKNUafLm2nB40mCt/PgJlijRP7YRvYN4m6trWZpa0NZ8Z8zrjtqeQbgQk
8RzpIt0JBcIUSD05QilLOoP9emLbC6vw0i6jL1yFe0QGY4rVxlSsKc0DNLBrKrSW
NIlxtDlRj9kT7aAFnReaVqR7ppAT8h8yVIHGp3R5DGhCzHwRyaksyBnhWSje03pN
BoeFqCC8LxDD9LQHODhDisGkDgAE8Kh398rdRB3u3yJChurNs9nIQoxacfmbxnNa
pA7YVZ5jRU89/yvtV92Why6FK067l1yibhVbOXIyfnG8jzlR/ar08QT212XBillx
uoJFUqBSj5EiAzbdOTryBg9M2k6Y9p5RHSYw2jowYs5UAOwHAS9jUb3M8a9lwA94
9xcgkMf7goPzfv1m7q6ym6PK6QeT24D30vgb0HDypOHs3KBgX6HIODsrkivGRKNX
UKGh7KFL4QqX+oTnXIGpJtUXsbdE5wbbnKqamk/QBnpOUP0OmCKaAm6cTlaANX5g
RL5wmfKZIJ1mc3SEi7EXGbvq94+lRzO8MkS24R5lelTsI33uLuBzpKIXdXC3oKqz
sPszqS9smCo+AvekXjoGupDy7FXqO1I0gGdffWj00Lyff8KKu7llhabLCUJMlGG2
FFK4sb/5geh7df45uPrmM7j7PI/WwCg0dbEwlprHMEqLtf/s5oIu654udafyNrpB
GLMsTSYKo6GEn5vV5KT2wXwjb6vpTwWZStrQT+0qc8RC/JWapw3VT0Un712/jJ7j
XvvVPzKM8ZNHG39Tp5FFuPnUBV8YgtKi7MdNsqOtTesE82DFwG1XbMRKyLERqe84
UIGPWbGablGrjk1FT75kqytJUi3awCd0UhYEhn9lolAI6QHeQvFhdX/2mPD1hZM8
nzkeck5nES3bYxxNDYxgZ3UOgk4yNIvwjzL3yfEZiProrgQUYOp8pi4zK+8Df8JD
DWOWgVXEa9T88uH+YVbkSm5PPS423xAuTM0kYSb0uW0uiu5YIODWZGrauz5JRzBJ
C7l687eKSIe1JaDy9xzJwuA+896DpGNjj4cFXaqPKMvTqz/VV7WKw+QWMtjfpiY3
lOjMrTkMX3BcFvQeLEB917crVrImKkWseF6kjKSQJY3g510VPoN6/KhanNfI84GG
fCX+mQ/0uvxa/DX7e92rBJ4931s94hyAqg1XgmAQlLS+WZL9aJJGl1OCukzcxJdV
T595jiBKNfh+F3AWlycp+HJiWRwf5Ex7BEoaJEuibcr+l6v0ugeH+YIJ45QKN/qb
yBBe005b2PCBTfM4Fz9SBPt3l7XZaWjwpUFkNej6RzbZpaY4A9wwa84Hlf5m9AHS
MoFmbdW/PeXRPSmR3u/b5EIcKnPdDX7T/Ul8VCLF0PyUw7IMEWX4IqAVV85G/dyf
poUo1Pcl84LEIxeGbs9R9b2tFNmeCNozFWqkDZV/D2jbqOWzWeUW8sA7PER98KSn
n0pqC0ydZz2dXJKzfDVnCvvV5QmMQZrI17pWtUga37bRKPQRN16q1+Kh7jXtRIGL
lAMZ9R5o8/RPpyePu2Iz6XuaPnSGd1GBFK2NGrZSBXb8ho4dM+Ncbval1il6yCzM
a7PbqQltDNmXWGPJ7nwU59fh3wXamZ+ipeDDDLADNeK2dTZDUc4JDAEMD0m0dT+5
UhGkX2tL5Ci3RyzQcjEfriPO6xkFThJyjwPUbPq6BqQa1rzwHTtsc8FBiroQ3xzt
y2+9W0cPM6OyfQ3JJscemBcUVER9TUELIEbWIGdV0g72w4N1sqHkBFO+LI3DTibA
QBJqG4jlAD52URfFIhtx2gdyhe3d5bt8kVjzvWbcZMUcv8IIA/bt9qO8QUJ3cwyo
gwDD9aDKkVbsQuiJ0+MO2h3u6A31qKdizRUPt69INS1ZGrFJFL6GMLnSqV+wgY0Q
+nbJhjNY+36ark/YcnnPJQWdrOzfKbqrKkVha01+hju66wmbYmym8rDUXMLSDpLw
9K4xX8JzYwhGVCxgv+WqAHdhdhmrE2H0R6pFYRAlFOyXJ6ZwB+V24f/HEWwZkwgD
Lzy39yXpvOTgUntyqDtDiDM5fP8W8e1dc8VdH9k78nw+lgSJaIMe6X6rYDf7JEJM
RrEwXYOpN5bGg0JLIxMTaHnTpJuy82InZCKwQ2neL/3FTCziz0vNX7offVrfUDCM
hCDZiEvJsFnlCdc3zZGHfoJJc0w4OJbFJk8loNV5bbegIpoWVE/7TwTNSNnfqCYj
aimvxnxMyK8NJAgsVYkb5oTvKBcPWTGy3SfQ9FEEkYVhOqZP0rVzhA+Qfx5v1dfo
mcc+IVs+TECNHPwB2+Y0foroJaDSYQc5Mq5HFP9y7zWF1351i/VauAnUgnQMwsv6
Ua5DqfGRKPgdePQS92tbd4RdPy6Rj0HESTAvaSsrC4wzAPCU1DdxD9NgqGBXA58V
9+TRJP252jedKm8rignYEYQHJM5Pv6gsMvRcvcUo3jVV4Ej6PqL+9sOkSwUsiOOB
FhT2fyJfY3XMU7X6hio+NadSvY9kICq6upIAjXqUT5Id1L0YgYdJBZu3w9Mj2irL
Ev82xD6HSlJgrX2QlPNnx3LRs+l50EQyy0SrshOpnZ9LWCtdO9gQ8hO+PM1X1sCY
auw+gSkk5k4XGM/zRelspfeb9596St/1jjSYaVs+bTouj9MMqkhy+t+tJdQqAl6I
IJJODMN6ziqH2Ua2xTHt6+6bBIf9KWgNkReqxH1J0sDHhdabKYkPBTSOLh6BA1eR
UrNzzwA9ukzvhiph9JUoyW4CzHLJI6o1lxjECAKft5y5gdqYumU37WXN2OSw7HeE
6wCj1QWbPbBdT1jE7SAebIXP4r8pyt5ivx7nSTN55N1D2H27jmIkPGt0v4SlbLw1
8rbk39t0aG3wWeowEbVGqra77MjTSLFl2/UqQyHugt/lh8Oc1VjfAxDitKszUqYZ
44NsUDWZsklNEEEjqJ9JnRpBzS/BMHIw+hG8BvbOhY8eUdEpqZyYEsMpuuipoN0x
pxCf1FyHs6Y8fE+6vlAymYGXHYI+8EQ2bOsVJWCqf/lCJD0+8jUCtQtjeex7DpWx
ycNA2dPHp5jL3tvDXKOg456SL5dz39Jl73Lc12975/vcUXnwIc6oHjnrhumyF24J
TyqIBPUC06kfjQ9lR6JoSK7l3hPL/PgVldMCMnHRmPQalCes1u3Ekx2sfdfn3WRZ
68PVefcyP345nra4kath/pFkGlaK/czwaw6z9zrBDmaVCG8f+zcVEXe9YihgMwMs
N4kgPCf778CZF2RHZ94IMv878dJGCrv8DQqqBiQaR03ZN7LZR4bheyNl50omjw85
7gCYIO5MoyAbCck1OlGMdd2kz4oVvidS+U5Uid2lgBXUtW0RyLHYd0lxcQnkxYP2
7CjvCFDHK7xOtrhFjs7pqLJUjQ1hbs4Kp0EQQQfj7AS8423FicN5P9f7qfpL73jM
QghyuetZE1GFECGYZmtXxXopWUu/njeXQvUyzXSFRoT/rQwWyvbfZ/4RkDufHP/e
xQMzLHJZlMUn0ggGvC4sTWBkJApnParc5VaEayCCFE0sgCiHnfxdf4AIxZgwBmUM
LaSImjte4UoUf8aE6f4Tidf3TYWXu6QyZqjgqBbNMWBQis3Naru3kn3yv04icdWv
hsNGryOyDJ3eHKLryv6kbb0bwCuI7ST040v5Hw+gXeeE5m+1vtB29P04tjVXg0+T
/6coAtg2ziLH0ZHlaCZNaqvFvFYfIhTSm1AOGvmj9awBg9/bdY6B1Wz3xGzK7+cr
BBy1MRKL72FVtzrJgFvDaleYTdL2ImVkcEfcNNcyUOLBE4QewB2oRxFfnsvRhmeU
2+oO9a7nHhgTFkgfc/qXVGe/EhWA6IyScFs28Ht7p+xuSnQTreu09gThluN5tcOQ
J2Dn5FaHeC2zaFCFaG/kM7a6nkakbfXyO/OSL8Ge/XjoAXyyllcsL8mveLtSw2Yr
w8LCwtz48haejMe67EZimx41lzdwVLBSitzq3wRnrdcYuaXZ/qs6Rz30xSDpzqNi
WbSCHjIt8EqV/WL6KRTqOxeIrz44JpPd2g2ejyrie1g9Ze+zvgN3yVuF8PUEe3iB
l4vKcLumr9+Tt0bQDKlxH2vkQcvXeGWd585Opf4XFAyMfUD8soAUICYQbKDi/I1g
4MXg54tnlriJbdKp+1premUeSBa4fbG9hDY/xmFfDn2MxBcUnpSX8RHu5tYG9doK
UZNCdn70WzZA0Sw8kyG/C63bPmV0ZiwvoBT5tXYfm2I5xDmZY/dNjr81M70k9Xll
+C0gIY8NXKgv9E/QEHl6+9KcC69+CBgkJ33nY2MeccT1YhDK94d0qlLI6eAoCn9c
4+pFWamE34YmCIPn4Im8CUop5Vg0NDe6ZD8DOuqxjfRnuTmCIJhVGCkUztmzP0Ic
rl2LuYXMn2m0slSyFEzZYw/gWKg7P9yaptioe3q8KMvURC615hKGX94=
=leE+
-----END PGP MESSAGE-----
```

**Comment by romanz** [comment]
Could you please update the `trezor-agent` to the latest git [commit](https://github.com/romanz/trezor-agent/commit/HEAD)?
```
$ git pull --ff-only
$ pip3 install --user -e .
```
IIUC, you may be running an older [version](https://github.com/romanz/trezor-agent/blob/385fc9457b0ac897259a4f7eea667e4133973547/libagent/device/trezor.py) (according to the logs line numbers)...

**Comment by romanz** [comment]
> gpg: signing failed: Broken pipe

It probably happened due to `'Unexpected magic characters'` error (from the log).
This may be a transient issue - could you please try again?

**Comment by TreyFiello** [comment]
@romanz OK - I did:
```
pip-autoremove trezor-agent -y
git clone https://github.com/romanz/trezor-agent
cd trezor-agent
pip install --user -e ./
```
Note: I had then had to do:
`pip install --user -e ./agents/trezor` for the trezor-agent binary to be on my path. Not sure if I installed this correctly or not.

Then followed your instructions from the last post, opened `trezor-agent` in a separate terminal and then:
`$ echo 123 | gpg2 --sign -sa`

This time the gpg2 command worked:
```
$ echo 123 | gpg2 --sign -sa
gpg: using "myemail@email.com" as default secret key for signing
-----BEGIN PGP MESSAGE-----

owGbwMvMwCW8jO/zsS2+bZcYT3MlMUT1W903NDLm6uhgYRDmYjAQU2RZYbI46s8L
RvlzZlFwtaxMIIVCMrmVqbmJmTkOYFIvOT+XgYtTAKboxWyG/5Wpr/ecuKChF87k
1XP05DEJG+Pr360ll37iWmZjWrty+1WGf3Yyz6uSp9kVPD39wnbvcYYJQn8Crjps
c4stKKvlPvS/6zkA
=Mdln
-----END PGP MESSAGE-----
```

... but in the trezor-agent window it prompted me for a passphrase (non-gui). If I type in my passphrase there the decryption succeeds. Otherwise it would just hang.

 Log below up to the point of it prompting me for a passphrase:
```
-----BEGIN PGP MESSAGE-----
Version: Keybase OpenPGP v2.0.76
Comment: https://keybase.io/crypto

wX4DT6XItd1c+h8SAgMEgzJ6yafu+BTMBsaTqFIKimToWo+eoEy4hLzL2RYE/uA3
5QlURn44Pjd8FhtMPicZScVbpGpcsgM2ozRgPgiPuDDqmptMZHnQvGPevY/W4idO
wsDqob0PfsXP9QoQmjmx8KIeC58SX5apc5ajQVT5ut3SyRYBARl3yBreeD3Cvu1M
6UTK+D8J2Xx7gWP6SubANeXi77vwyKGfcqMcIsBZjBvQeMk/OUml7woULIqlIBz8
oFdufSaCKoKNJHuz1Q/zzEK28EWZoLVGXm6mq53j+Ihn/LP3wV4sUx29b/EHclFq
Zx37Ox1l9PnG2OnAV1f2GSt9wwyIzOCELy9aoaIOJVrjB2kFPogPTs/2vLx38ax+
EHQtcNHN+oK48KzhRm1Ja2VXTPFB1EWMwuUN3XkDATyy9GQIva+09fRYSW7sa4DU
VfSZOLCIrfH/z5tCIMOItCV6i3/q83RO32LnV7IIJoIwryIfojAfFgRsAQJwK9l3
tiVzx4u94IIgIHyw7NH2xMQC+hSJIcKoAL4DwsnXGaScn7RQ+VkcOiOQcoi0UgHq
d3Nic7FgxXSHJEtFsNfvjDbA4YugC+JkAlAyelqFSroMUj9uZvHEYK7n1R503k3x
TimLr0j+BUtjn2J4U02tFxElVUF5Kgbh1xUFRhCwfb10pJUi0zzA4sXsJIq7QoxL
COBCtaRJtTSNZEU58Trq6NS9LsHx8xoy8F4mV47YMG9VqusO0VNzQv30wKYfDNnP
kEd4TRMg/9oGkxH0haiPnDDwZ7rN+EYRxPLyDSS+oo/PFhj00zAHZCFXJ2cJCBGx
5PNHKr/O3dIEwxgXaxezIiDhlFNQylFme2wNoQYZsNSD4cUssHDiL6uv49DG0LsT
G/MoUha5v3UX6nblrBdSuz0YxZ/jPbqcGD7IsQ1bi5ElZ4NME/RLFCGEHAylI1Be
Ebovn+NWQjbneLHyC6zE3cugJHlLdN+eZGdTLTsRFJ/btjSwjVEPZUwqrUlZZ5Pk
lmTmWWJjtlbNJso3JNx+KvXwL3pBUcSXQo+fB74Ag12xQsbz7UC1VgIySgZRyuDq
XrBfCyA0r0v7WNbAHeImPs0yBBk3fiTRo9mLLcNJNA0qwlSGScDi/mc9vKantWkh
qGcD8cxZciA9V3SY3x0pwC7osOtIT9ttK1T9GQ9TVLd4OadkQWCZCkCjoVua9Jbl
xMmPJPAaUs8azl9JgbJhnWCVD1ycR+j2ZC5h45/FyURYkJhJx6iu8Nt9tRkOGgkI
y1prslTa46cNHyz9LB1KPlOUWhk/R4rF0gI9Y5SXuRnb/P1ZNN3AymCV77T4uTyY
S6/HyZUCg3YgELLnZMRZ5xH202fanlfQNlfTM7zE1KqiIgPQ9TTczxSxZRRB/2Ve
Oex24pfYx/tlYxpXzrXJJF/A4xOn2nkafnc+WEHytwK50AreM//KNI9l4Q0JkcUn
lzhPoKVACoWaT7cIFwUc5zygcXdqn52FkyW2QnwmcKbeOuZDhQKBpeUsVtmBdrOC
xnrjjfuCJA3iYc/8EHfgGW8FOv1+syTuwpExlYTUOr7Zxzlc/biT/1kkxnz7+cZu
65/X75mt+fdve03Ok39pcFD/fHwsHeo4NahCk3ml6pMj37cs35ezcGE3DOj/682H
JdLKX8cQkrHXjVJ6dZTkE61NCcCwYlCgBPULqv29aq909qtVEqa6rtR2WYXGDR7T
x/Bu007L+pgusfWlCcrEhZNczwG6vefeDD4G/6uFCgOrs/nOEWzz8qeBgdgoVaeM
aPRmRXnaErd+bCiFjtZaBv/c6+OZkZn4C39rRK4ux0adaKPs0Zu22wkc3zhWK7Wz
puS9yHFHL/DJnrT6Pk0bFoYPONAWVSf8S1SUF7HqABiyxd2ByRZj24nrKXhwSNxt
cq2CkbLirL9hEzGzT1FrFk5QyAKCtxGu9jfqsVb8o7fb4ra1b5gEmyEnML9lUrZK
KYopoB08j2BQfsmSGbVu1ntqNIjXCSgLD7/VIYBqE0w2Zu3RPrpvjHBgQna7Y3Un
sdfI99Gk6iO0CvJvra2G/NdCX3+X/NLLMdy4ITFNzsPxALKTRlEOOG1qdSAzn5Zz
IPm+BAqBlsZmSykdewowFVJDRjO3pCYYxo3zoiGUJp8nfz4/EuaUP3+AEo8hMObI
hReraVp064b1gRo7AtFGdWK524g+1nWofm5CkW48eaX+mvXscCzFcKhZMcTArAOG
pDwxNavOH9SK6Kn2xr0PGHPCCwHn8vDHfoPoeUV+3nl6Zd3TFoG9bzOf9+lP+8ze
Tniif5R1q+EC8UAIRJPIYFwGeY1zeOoU8CdDWhaMiDWerZPqaeiyfZC29Ls4PoD1
VwJ5FyMn5A/cYiOWOOYMYPg1K3/FuZTj0BP2PEhb2h5YvxkCgtiuze29XqPVuXEg
q9HjXJLGP9Ag3GrF7PSiwZocNddwAFXZ/YPE1DZ0VWeTxKxHqFZ2RBCJfedewil7
e+LxQaUeYA++0PDHDMQNErZw+efk84VimJL/VdqlOvtDXjoQ78yAL1NOcUy4nR1c
9+2P0kK/xT9eNH3er37j5rMU7Nbc4Pet9se1XA0bDnRYodpCaz5kMS8LCXfo9Dw7
TxhojzKkf65F1MqW3GkZANRekcWaS5Led83MRkNIIeeuCzggxP9SL628XveCrAz6
HupxgcAIIJYO2YzcSOCVeV3BPxegENo3r0tMUzLjOVsUWqRu+JjLKXY9gQlQYvpI
6xyHIcL8pwMdZS3TndfmgTRFYGepDLub2eKgvnGhtx5OAcQRTPmoy+nykb2bRsru
MWo3Oo/UDpwoBYJn2sU6HPofdjZNUDUBEMbRf1dJ6mAH4L6cgHGUo2VibGxP1nPN
jaeACoCQRi31IeQmwDw1jUhU1mWlA/cqUfSfWQRoeONGA0D5P5YD4zTrcvbH6LiG
XEJC5N2OCUYniCvlgzoMeom5h2UqWIruvhmTPrbnoV1QVjcuVOqcQou8j+RVks9v
vWsP48+YbOWeNaigRBd+/NZ1IolBiJozt1oAbfAr35US3BqBjn3OuRjyezQlgkDU
P8l0UB3JjibTbEnBwlMn3os7ZUvi0XtxJaDKvBlrhlZTSf91UATbP8Yeizjuvo4j
6E3caVNCjWGG10iuXffdD6P1z0r/hZBP7yc0hhH9J0r7DvTvGKf2JNRQ4U9MvkK1
qHuGS2YDkkgKnLc11c1sR4FR1SXpmruSbS96/eahwVCBLbGil+HP1Z3y98j7XFjG
TcOVWv3LN+fY5jUfhmzIuN17FaQN/28bWBFi/t0DL27siX7ld3Lu0Z2Nwbvexcen
GtjPoJIFFZFDb6OesRoiZFyfRqiv+EM56SHUSs2S9H5JKwxM0bGQJla+Mgaey6O+
VaJEn9gy8JKminBzB+5S7exaAKCly0D+JNSB1lvy5eho4LR9xo9Gcldot5OVnYlX
oyGxbonHID3X2nEjLjmnEVgaop2XTIw2jWFPBdMprIAYfzEDCkt9wWuchjupMu9I
dzETOIKgwtj3
=VtO6
-----END PGP MESSAGE-----
```

**Comment by romanz** [comment]
In case `trezor-gpg-agent` is running under an active terminal https://github.com/romanz/trezor-agent/blob/654a3c465a053175d1ca7e59f506d87e8b9ffca1/libagent/device/trezor.py#L93
we use the console-based passphrase entry, instead of the tk-based UI:
https://github.com/romanz/trezor-agent/blob/654a3c465a053175d1ca7e59f506d87e8b9ffca1/libagent/device/trezor.py#L17

IIUC, the following should now work correctly (i.e. invoke the UI-base passphrase entry):
```
$ killall trezor-gpg-agent
$ echo 123 | gpg2 --sign -a  # should start a separate trezor-gpg-agent process in the background.
```

**Comment by TreyFiello** [comment]
@romanz confirmed working and fixed!!!! Both your sign command above and  `gpg2 -d test.txt.gpg` now do not prompt for passphrase.

To those trying to get this working you must follow @romanz's note above:

> I can suggest adding the following command to the ~/.gnupg/trezor/run-agent.sh file:
> 
> #!/bin/sh
> export PATH=<snip>
> export PASSPHRASE='abc'
> /home/roman/.local/bin/trezor-gpg-agent $*

Will the passphrase environmental variable also work for SSH keys? If not, would you like me to open up a separate issue? Thank you for your assistance!

**Comment by romanz** [comment]
You're welcome :)

> Will the passphrase environmental variable also work for SSH keys?

It should work as well, since it uses the same [code](https://github.com/romanz/trezor-agent/blob/master/libagent/device/trezor.py) to communicate with the hardware device.

**Comment by TreyFiello** [comment]
@romanz I have been using this for a couple days and I have noticed 1 issue: 

When I first restart my computer I am still prompted for the passphrase via the GUI popup box 1 time (after entering it this first time I am never prompted again).

I tailed gpg-agent.log and found it is popping up at this location:
```
2018-02-28 08:59:08,283 DEBUG bip32 address string: '\x00\x00\x00\x00gpg://myemail@email.com' [interface.py:88]
```

If I hit cancel to entering my passphrase I see it error out here:

```
 2018-02-28 09:02:51,618 ERROR        UI failed: 'Traceback (most recent call last):\n  File "<string>", line 1, in <module>\nTypeError: expected a string or other character buffer object\n' [trezor.py:26]
 2018-02-28 09:02:51,815 ERROR        handler failed: Command '['/usr/local/opt/python/bin/python2.7', '-c', 'import sys, pymsgbox; sys.stdout.write(pymsgbox.password(sys.stdin.read()))']' returned non-zero exit status >
  Traceback (most recent call last):
   File "/Users/test/trezor-agent/libagent/gpg/__init__.py", line 231, in run_agent
     handler.handle(conn)

```

**Comment by romanz** [comment]
Yes, this happens (unfortunately) because I have added a simple passphrase cache in trezor-gpg-agent (https://github.com/romanz/trezor-agent/commit/73d60dbec0c1516577aa4e7a406bd20bf6458e27), so you currently need to re-type the passphrase after a restart :(

**Comment by TreyFiello** [comment]
@romanz oh ok I didn't know we completely went away from the environmental variable.

**Comment by TreyFiello** [comment]
@romanz I noticed when I was SSHing the passphrase environmental variable was still being respected, (you may already know this) - but just FYI I found out the official python-trezor library uses the PASSPHRASE environmental variable just in case you wanted to stay consistant:

https://github.com/trezor/python-trezor/blob/master/trezorlib/client.py#L289-L292
